aibolit/metrics/fanout/FanOut.py

```diff
@@ -1,66 +1,112 @@
-from aibolit.utils.ast_builder import build_ast
-from aibolit.ast_framework import AST, ASTNodeType
-from itertools import islice
-from aibolit.ast_framework.java_package import JavaPackage
+# The MIT License (MIT)
+#
+# Copyright (c) 2020 Aibolit
+#
+# Permission is hereby granted, free of charge, to any person obtaining a copy
+# of this software and associated documentation files (the "Software"), to deal
+# in the Software without restriction, including without limitation the rights
+# to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+# copies of the Software, and to permit persons to whom the Software is
+# furnished to do so, subject to the following conditions:
+#
+# The above copyright notice and this permission notice shall be included
+# in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+# FITNESS FOR A PARTICULAR PURPOSE AND NON-INFRINGEMENT. IN NO EVENT SHALL THE
+# AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+# OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+# SOFTWARE.
+
+from typing import Set
+
+from aibolit.ast_framework import AST, ASTNode, ASTNodeType
 
 
 class FanOut:
-    '''
+    """
     Fan Out metric is defined as the number of other classes referenced by a class.
-    '''
-    def __init__(self):
-        pass
+    """
+
+    def value(self, ast: AST) -> int:
+        fan_out = 0
+        for class_declaration in ast.get_subtrees(ASTNodeType.CLASS_DECLARATION):
+            fan_out += self._calculate_class_fan_out(class_declaration)
+
+        return fan_out
+
+    def _calculate_class_fan_out(self, java_class: AST) -> int:
+        class_declaration = java_class.get_root()
+        assert class_declaration.node_type == ASTNodeType.CLASS_DECLARATION
```

Comment:
Don't you think that this assert is EXTRA? In function value in cycle FOR you iterate through each node with type `ASTNodeType.CLASS_DECLARATION`. After this node goes to the function `_calculate_class_fan_out`. What else can goes wrong?

Reply:
In current version of code this `assert` will never raise an exception. But later on, the method, calling this one may change. I want to guarantee, that further code of this method may rely on some solid prerequisites. It is like additional type checking. Because of dynamic nature of `ASTNode` we can't guarantee, weather or not this field is available with mypy. But with such `asssert` we get this guarantee. It is a defense programming, when  your function trust no one and tries to check input parameters as hard as possible.